tests/devices/unit_tests/current_amplifier/test_SR570.py

```diff
@@ -0,0 +1,280 @@
+    assert (
+        sleep.call_count
+        == int(SR570GainToCurrentTable(final_gain).name.split("_")[-1])
+        - int(SR570GainToCurrentTable(starting_gain).name.split("_")[-1])
+        + 1
+    )
+
+
+@pytest.mark.parametrize(
+    "starting_gain, gain_change_count, final_gain",
+    [
+        ([1e3, 5, 1e3]),
+        ([5e4, 2, 1e4]),
+        ([2e5, 3, 2e4]),
+        ([1e12, 13, 5e7]),
+        ([1e5, 20, 1e3]),
+    ],
+)
+@mock.patch("asyncio.sleep")
+async def test_SR570_decrease_gain(
+    sleep: AsyncMock,
+    mock_sr570: SR570,
+    starting_gain: str,
+    gain_change_count: int,
+    final_gain: str,
+):
+    await mock_sr570.set(starting_gain)
+    for _ in range(gain_change_count):
+        await mock_sr570.decrease_gain()
+    assert (await mock_sr570.get_gain()) == SR570GainToCurrentTable(final_gain).name
+    assert (
+        sleep.call_count
+        == abs(
+            int(SR570GainToCurrentTable(final_gain).name.split("_")[-1])
+            - int(SR570GainToCurrentTable(starting_gain).name.split("_")[-1])
+        )
+        + 1  # for starting gain
+    )
+
+
+@pytest.mark.parametrize(
+    "gain,raw_count, expected_current",
+    [
+        ("sen_1", 0.51e5, 0.51e-3),
+        ("sen_3", -10e5, -2e-3),
+        ("sen_11", 5.2e5, 2.6e-6),
+        ("sen_17", 2.2e5, 1.1e-8),
+        ("sen_23", 8.7e5, 4.35e-10),
+        ("sen_5", 0.0, 0.0),
+    ],
+)
+async def test_SR570_struck_scaler_read(
+    mock_sr570: SR570,
+    mock_sr570_struck_scaler_detector,
+    RE: RunEngine,
+    gain,
+    raw_count,
+    expected_current,
+):
+    temp = SR570FullGainTable[gain].value
+    set_mock_value(mock_sr570.coarse_gain, temp[0])
+    set_mock_value(mock_sr570.fine_gain, temp[1])
+    set_mock_value(mock_sr570_struck_scaler_detector.counter().readout, raw_count)
+    set_mock_value(mock_sr570_struck_scaler_detector.counter().count_time, 1)
+    set_mock_value(mock_sr570_struck_scaler_detector.auto_mode, False)
+    docs = defaultdict(list)
+
+    def capture_emitted(name, doc):
+        docs[name].append(doc)
+
+    RE(count([mock_sr570_struck_scaler_detector]), capture_emitted)
+    assert docs["event"][0]["data"][
+        "mock_sr570_struck_scaler_detector-current"
+    ] == pytest.approx(expected_current)
+
+
+@pytest.mark.parametrize(
+    "gain,raw_count, expected_current",
+    [
+        (
+            "sen_10",
+            [1e2, 2e2, 5e2, 1e3, 2e3, 5e3, 1e4, 2e4, 5e4, 5e4],
+            1e-9,
+        ),
+        ("sen_1", [1e5, 1e5], 1e-3),
+        (
+            "sen_10",
+            [520e5, 260e5, 104e5, 52e5, 26e5, 10.4e5, 5.2e5, 2.6e5, 2.6e5],
+            5.2e-4,
+        ),
+        ("sen_19", [22.2e5, 11.1e5, 4.4e5, 4.4e5], 2.2e-8),
+        ("sen_5", [0.0] * (30 - 5), 0.0),
+        ("sen_5", [-200.0e5, -100.0e5, -50.0e5, -20e5, -10e5, -10e5], -0.01),
+        ("sen_25", [0.002e5, 0.004e5, 0.01e5, 0.02e5, 0.02e5], 2e-14),
+    ],
+)
+async def test_SR570_struck_scaler_read_with_autoGain(
```

Comment:
Must: In dodal we consider it very important to keep the total time required to run the test suite to a minimum (so that developers actually do it regularly), which means you should not write long-running tests, there are a few tests in this file and `test_femto.py` that seem to take a long time.

See the following snippet from running `tox -e tests -- --durations=10 tests/devices/unit_tests/current_amplifier`:

```
========================================================================================================================================= slowest 10 durations ==========================================================================================================================================
3.35s call     tests/devices/unit_tests/current_amplifier/test_SR570.py::test_SR570_struck_scaler_read_with_autoGain[sen_5-raw_count4-0.0]
1.28s call     tests/devices/unit_tests/current_amplifier/test_SR570.py::test_SR570_struck_scaler_read_with_autoGain[sen_10-raw_count0-1e-09]
0.86s call     tests/devices/unit_tests/current_amplifier/test_SR570.py::test_SR570_struck_scaler_read_with_autoGain[sen_25-raw_count6-2e-14]
0.79s call     tests/devices/unit_tests/current_amplifier/test_femto.py::test_femto_struck_scaler_read_with_autoGain[sen_1-raw_voltage7-1e-23]
0.79s call     tests/devices/unit_tests/current_amplifier/test_femto.py::test_femto_struck_scaler_read_with_autoGain[sen_5-raw_voltage5-0.0]
0.50s call     tests/devices/unit_tests/current_amplifier/test_SR570.py::test_SR570_struck_scaler_read_with_autoGain[sen_19-raw_count3-2.2e-08]
0.43s call     tests/devices/unit_tests/current_amplifier/test_femto.py::test_femto_struck_scaler_read_with_autoGain[sen_10-raw_voltage0-1e-09]
0.13s call     tests/devices/unit_tests/current_amplifier/test_SR570.py::test_SR570_struck_scaler_read_with_autoGain[sen_10-raw_count2-0.00052]
0.09s call     tests/devices/unit_tests/current_amplifier/test_SR570.py::test_SR570_struck_scaler_read_with_autoGain[sen_5-raw_count5--0.01]
0.09s call     tests/devices/unit_tests/current_amplifier/test_SR570.py::test_sr570_set[2000000000.0-0.2-sen_20]
```

This is the longest-running test but this comment is a blanket for it and a couple of others. One way to slow them down is to split them into multiple tests with more mocking, but I'd need to take a closer look to see what is needed here.